File: server/app/models/agency.py
from datetime import datetime
from typing import Optional, Literal
from uuid import UUID
from pydantic import BaseModel, EmailStr


# Agency types
AgencyType = Literal["talent", "brand", "hybrid"]
VerificationStatus = Literal["pending", "in_review", "verified", "rejected"]
MemberRole = Literal["owner", "admin", "member"]


# Agency models
class AgencyCreate(BaseModel):
    name: str
    agency_type: AgencyType
    description: Optional[str] = None
    website_url: Optional[str] = None
    contact_email: Optional[EmailStr] = None
    industries: Optional[list[str]] = None


class AgencyUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    logo_url: Optional[str] = None
    website_url: Optional[str] = None
    contact_email: Optional[EmailStr] = None
    industries: Optional[list[str]] = None


class AgencyResponse(BaseModel):
    id: UUID
    name: str
    slug: str
    agency_type: AgencyType
    description: Optional[str]
    logo_url: Optional[str]
    website_url: Optional[str]
    is_verified: bool
    verification_status: VerificationStatus
    contact_email: Optional[str]
    industries: list[str]
    created_at: datetime
    updated_at: datetime


class AgencyPublicResponse(BaseModel):
    """Public agency profile for discovery."""
    id: UUID
    name: str
    slug: str
    agency_type: AgencyType
    description: Optional[str]
    logo_url: Optional[str]
    website_url: Optional[str]
    is_verified: bool
    industries: list[str]


# Member models
class AgencyMemberInvite(BaseModel):
    email: EmailStr
    role: MemberRole = "member"
    title: Optional[str] = None


class AgencyMemberUpdate(BaseModel):
    role: Optional[MemberRole] = None
    title: Optional[str] = None
    is_active: Optional[bool] = None


class AgencyMemberResponse(BaseModel):
    id: UUID
    agency_id: UUID
    user_id: UUID
    email: str
    role: MemberRole
    title: Optional[str]
    is_active: bool
    invited_at: datetime
    joined_at: Optional[datetime]


# Agency with member context
class AgencyWithMembership(BaseModel):
    agency: AgencyResponse
    membership: AgencyMemberResponse
    member_count: int
    active_deals_count: int

File: client/src/types/agency.ts
// Agency types
export type AgencyType = 'talent' | 'brand' | 'hybrid';
export type VerificationStatus = 'pending' | 'in_review' | 'verified' | 'rejected';
export type MemberRole = 'owner' | 'admin' | 'member';

// Agency
export interface Agency {
  id: string;
  name: string;
  slug: string;
  agency_type: AgencyType;
  description: string | null;
  logo_url: string | null;
  website_url: string | null;
  is_verified: boolean;
  verification_status: VerificationStatus;
  contact_email: string | null;
  industries: string[];
  created_at: string;
  updated_at: string;
}

export interface AgencyCreate {
  name: string;
  agency_type: AgencyType;
  description?: string;
  website_url?: string;
  contact_email?: string;
  industries?: string[];
}

export interface AgencyUpdate {
  name?: string;
  description?: string;
  logo_url?: string;
  website_url?: string;
  contact_email?: string;
  industries?: string[];
}

export interface AgencyPublic {
  id: string;
  name: string;
  slug: string;
  agency_type: AgencyType;
  description: string | null;
  logo_url: string | null;
  website_url: string | null;
  is_verified: boolean;
  industries: string[];
}

// Agency Member
export interface AgencyMember {
  id: string;
  agency_id: string;
  user_id: string;
  email: string;
  role: MemberRole;
  title: string | null;
  is_active: boolean;
  invited_at: string;
  joined_at: string | null;
}

export interface AgencyMemberInvite {
  email: string;
  role?: MemberRole;
  title?: string;
}

export interface AgencyMemberUpdate {
  role?: MemberRole;
  title?: string;
  is_active?: boolean;
}

// Agency with membership context
export interface AgencyWithMembership {
  agency: Agency;
  membership: AgencyMember;
  member_count: number;
  active_deals_count: number;
}

// Registration
export interface AgencyRegister {
  email: string;
  password: string;
  agency_name: string;
  agency_type: AgencyType;
  description?: string;
  website_url?: string;
  industries?: string[];
}
